File: packages/Teams/dotnet/TriggerConditions/OnTeamsMeetingStart.cs
// Licensed under the MIT License.
// Copyright (c) Microsoft Corporation. All rights reserved.

using System.Collections.Generic;
using System.Runtime.CompilerServices;
using AdaptiveExpressions;
using Microsoft.Bot.Builder.Dialogs;
using Microsoft.Bot.Builder.Dialogs.Adaptive.Conditions;
using Microsoft.Bot.Connector;
using Newtonsoft.Json;

namespace Microsoft.Bot.Components.Teams.Conditions
{
    /// <summary>
    /// Actions triggered when a Teams Meeting Start event is received.
    /// </summary>
    /// <remarks>
    /// turn.activity.value has meeting data.
    /// </remarks>
    public class OnTeamsMeetingStart : OnEventActivity
    {
        [JsonProperty("$kind")]
        public new const string Kind = "Teams.OnMeetingStart";

        [JsonConstructor]
        public OnTeamsMeetingStart(List<Dialog> actions = null, string condition = null, [CallerFilePath] string callerPath = "", [CallerLineNumber] int callerLine = 0)
            : base(actions: actions, condition: condition, callerPath: callerPath, callerLine: callerLine)
        {
        }

        /// <inheritdoc/>
        protected override Expression CreateExpression()
        {
            return Expression.AndExpression(Expression.Parse($"{TurnPath.Activity}.channelId == '{Channels.Msteams}' && {TurnPath.Activity}.name == 'application/vnd.microsoft.meetingStart'"), base.CreateExpression());
        }
    }
}

File: packages/Teams/js/src/conditions/onTeamsMeetingStart.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { Expression } from 'adaptive-expressions';
import { Channels } from 'botbuilder';
import { TurnPath } from 'botbuilder-dialogs';
import { OnEventActivity } from 'botbuilder-dialogs-adaptive';

/**
 * Actions triggered when a Teams Meeting Start event is received.
 * Note: turn.activity.value has meeting data.
 */
export class OnTeamsMeetingStart extends OnEventActivity {
  static $kind = 'Teams.OnMeetingStart';

  /**
   * Create expression for this condition.
   *
   * @returns {Expression} An [Expression](xref:adaptive-expressions.Expression) used to evaluate this rule.
   */
  protected createExpression(): Expression {
    return Expression.andExpression(
      Expression.parse(
        `${TurnPath.activity}.channelId == '${Channels.Msteams}' && ${TurnPath.activity}.name == 'application/vnd.microsoft.meetingStart'`
      ),
      super.createExpression()
    );
  }
}
